docs/dev/datastream_execution_mode.md

```diff
@@ -0,0 +1,377 @@
+request the current processing time and register processing time timers, but,
+as in the case of Event Time, all the timers are going to fire at the end of
+the input.
+
+Conceptually, we can imagine that processing time does not advance during the
+execution of a job and we fast-forward to the *end of time* when the whole
+input is processed.
+
+### Failure Recovery
+
+In `STREAMING` execution mode, Flink uses checkpoints for failure recovery.
+Take a look at the [checkpointing documentation]({% link
+dev/stream/state/checkpointing.md %}) for hands-on documentation about this and
+how to configure it. There is also a more introductory section about [fault
+tolerance via state snapshots]({% link learn-flink/fault_tolerance.md %}) that
+explains the concepts at a higher level.
+
+One of the characteristics of checkpointing for failure recovery is that Flink
+will restart all the running tasks from a checkpoint in case of a failure. This
+can be more costly than what we have to do in `BATCH` mode (as explained
+below), which is one of the reasons that you should use `BATCH` execution mode
+if your job allows it.
+
+In `BATCH` execution mode, Flink will try and backtrack to previous processing
+stages for which intermediate results are still available. Potentially, only
+the tasks that failed (or their predecessors in the graph) will have to be
+restarted, which can improve processing efficiency and overall processing time
+of the job compared to restarting all tasks from a checkpoint.
+
+## Important Considerations
+
+Compared to classic `STREAMING` execution mode, in `BATCH` mode some things
+might not work as expected. Some features will work slightly differently while
+others are not supported.
+
+Behavior Change in BATCH mode:
+
+* "Rolling" operations such as [reduce()]({% link dev/stream/operators/index.md
+  %}#reduce) or [sum()]({% link dev/stream/operators/index.md %}#aggregations)
+  emit an incremental update for every new record that arrives in `STREAMING`
+  mode. In `BATCH` mode, these operations are not "rolling". They emit only the
+  final result.
+
+
+Unsupported in BATCH mode:
+
+* [Checkpointing]({% link concepts/stateful-stream-processing.md
+  %}#stateful-stream-processing) and any operations that depend on
+  checkpointing do not work.
+* [Broadcast State]({% link dev/stream/state/broadcast_state.md %})
+* [Iterations]({% link dev/stream/operators/index.md %}#iterate)
+
+Custom operators should be implemented with care, otherwise they might behave
+improperly. See also additional explanations below for more details.
+
+### Checkpointing
+
+As explained [above](#failure-recovery), fault tolerance for batch programs
+does not use checkpointing. Recovery happens by fully replaying the streams.
+This is possible because inputs are bounded.  This pushes the cost more towards
+the recovery, but makes the regular processing cheaper, because it avoids
+checkpoints.
```

Comment:
The last paragraph of `failure recovery` reads as if `BATCH` execution improves the overall execution time of jobs but here it reads a bit differently. Concretely that `BATCH` recoveries are more costly than `STREAMING` recoveries.

Reply:
Good catch! @dawidwys what was the original intention here?

Reply:
My intention was to briefly remind the batch failure recovery model. For that I actually reused the description from: https://ci.apache.org/projects/flink/flink-docs-master/concepts/stateful-stream-processing.html#state-and-fault-tolerance-in-batch-programs

Reply:
With the description in the `failure recovery` section, we can probably drop the first paragraph and start with the second one:

>It is important to remember that because there are no checkpoints, as described [above](#failure-recovery), certain ...

BTW shall we update it in `#stateful-stream-processing.html#state-and-fault-tolerance-in-batch-programs`? It is not easy to tell which model recovers "faster" as it very much depends on the state size, number of records to replay, number of tasks to recover etc.

Reply:
Oh boy, this text was first added in 2016: https://github.com/apache/flink/blame/b04d51a129a3341887e7a0866557c9871f58e94c/docs/concepts/concepts.md. I copied it to the current concepts section from there. That's not at all up-to-date anymore.

That section needs an overhaul or should be removed because it's also misleading for `DataSet` programs or Table/SQL batch programs.

Reply:
For this documentation here I think we can go with @dawidwys's suggestion and just drop that paragraph because I added some text about that above.